elastic/kibana
# [Upgrade Assistant] Fix potentially long URI

## Summary

When a cluster has a lot of indices (1000s) with deprecations the Upgrade Assistant can build up a really long URI to check whether they are closed. This can cause 414s to be returned by web servers. Even though, technically, URIs do not have length limits, web servers' default configuration will limit the maximum length of the URI and cause Kibana's check to be rejected before it reaches ES.

The proposed fix is to just query `*` indices which is more expensive, but avoids running into the URI length blocker.

Fixes https://github.com/elastic/kibana/issues/72616

## How to reproduce

Ensure that you are on current master. This is a more hacky way of reproducing the error.

In `x-pack/plugins/upgrade_assistant/server/lib/es_migration_apis.ts`, replace line 23 with:

```ts
  let indices: any[] = [];
  for (let x = 0; x < 1000; ++x) {
    indices = indices.concat([{ index: 'test1' }, { index: 'test2' }, { index: 'test3' }]);
  }
```

Open your browser debug console and view the network tab and go to Upgrade Assistant. You should now see the error `[too_long_frame_exception] An HTTP line is larger than 4096 bytes.` reported as a 500 response.

## How to test

This is slightly tricky to test on latest master without actual deprecated indices that can be opened and closed.

The hack route for testing the logic changes with actual ES response:

First replace line 43 of `x-pack/plugins/upgrade_assistant/server/lib/es_migration_apis.ts` with fake deprecation data:

<details>
<summary>Fake data</summary>

```
 const fakeDeprecations = [
    {
      level: 'warning',
      message: 'none level test',
      url: 'test/url',
      details: 'test details',
      index: 'test1',
      reindex: true,
      needsDefaultFields: false,
      blockerForReindexing: 'index-closed', // Faking that test1 is closed!
    },
    {
      level: 'warning',
      message: 'info level test',
      url: 'test/url',
      details: 'test details',
      index: 'test2',
      reindex: true,
      needsDefaultFields: false,
    },
    {
      level: 'warning',
      message: 'warning level test2',
      url: 'test/url',
      details: 'test details',
      index: 'test3',
      reindex: true,
      needsDefaultFields: false,
    },
    {
      level: 'critical',
      message: 'metricbeat-1 successful fix',
      url: 'test/url',
      details: 'test details',
      index: 'metricbeat-1',
      reindex: false,
      needsDefaultFields: true,
    },
    {
      level: 'critical',
      message: 'metricbeat-2 failed fix',
      url: 'test/url',
      details: 'test details',
      index: 'metricbeat-2',
      reindex: false,
      needsDefaultFields: true,
    },
  ];
  return {
    readyForUpgrade: criticalWarnings.length === 0,
    cluster: cluster.concat(fakeDeprecations),
    indices: indices.concat(fakeDeprecations),
  };
```
</details>

Then do the following to test that we are correctly handling the new resolve indices response to check that index is open or closed:

1. Inside of `x-pack/plugins/upgrade_assistant/server/lib/es_migration_apis.ts` replace line 23 with the snippet from above.
2. Then just after line 37 add a `console.log(indexData)`.
3. In dev tools console run `PUT test1` (do the same for 2 & 3)
4. Start Kibana and go to the Upgrade Assistant
6. In the console output of Kibana you should see something like three empty objects
7. Close test1 in console (`POST test1/_close`)
8. Go to the Upgrade Assistant again. This time the first log should state `{ blockerForReindexing: 'index-closed' }`.
